spring/src/main/java/com/linecorp/armeria/spring/ArmeriaAutoConfiguration.java

```diff
@@ -0,0 +1,208 @@
+import com.linecorp.armeria.common.http.HttpStatus;
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.docs.DocServiceBuilder;
+import com.linecorp.armeria.server.http.AbstractHttpService;
+import com.linecorp.armeria.server.http.healthcheck.HealthChecker;
+import com.linecorp.armeria.server.http.healthcheck.HttpHealthCheckService;
+import com.linecorp.armeria.server.logging.DropwizardMetricCollectingService;
+import com.linecorp.armeria.spring.ArmeriaSettings.Port;
+
+import io.netty.util.NetUtil;
+
+@Configuration
+@EnableConfigurationProperties(ArmeriaSettings.class)
+@ConditionalOnMissingBean(Server.class)
+public class ArmeriaAutoConfiguration {
+
+    private static final Logger logger = LoggerFactory.getLogger(ArmeriaAutoConfiguration.class);
+
+    private static final HealthChecker[] EMPTY_HEALTH_CHECKERS = new HealthChecker[0];
+
+    private static final Port DEFAULT_PORT = new Port().setPort(8080)
+                                                       .setProtocol(HttpSessionProtocols.HTTP);
+
+    /**
+     * Create a {@link Server} bean.
+     */
+    @Bean
+    public Server armeriaServer(
+            ArmeriaSettings armeriaSettings,
+            MetricRegistry metricRegistry,
+            Optional<List<HealthChecker>> healthCheckers,
+            Optional<List<ArmeriaServiceInitializer>> armeriaServiceInitializers,
+            Optional<List<ThriftServiceRegistrationBean>> thriftServiceRegistrationBeans,
+            Optional<List<HttpServiceRegistrationBean>> httpServiceRegistrationBeans)
+            throws InterruptedException {
+        if (!armeriaServiceInitializers.isPresent() &&
+            !thriftServiceRegistrationBeans.isPresent() &&
+            !httpServiceRegistrationBeans.isPresent()) {
+            // No services to register, no need to start up armeria server.
+            return null;
+        }
+
+        ServerBuilder server = new ServerBuilder()
+                .gracefulShutdownTimeout(
+                        armeriaSettings.getGracefulShutdownQuietPeriodMillis(),
+                        armeriaSettings.getGracefulShutdownTimeoutMillis());
+
+        configurePorts(armeriaSettings, server);
+
+        List<TBase<?, ?>> docServiceRequests = new ArrayList<>();
+        thriftServiceRegistrationBeans.ifPresent(beans -> beans.forEach(bean -> {
+            @SuppressWarnings("unchecked")
+            Service<HttpRequest, HttpResponse> service =
+                    (Service<HttpRequest, HttpResponse>) bean.getService();
+
+            service = service.decorate(
+                    DropwizardMetricCollectingService.newDecorator(
+                            metricRegistry, serviceMetricName(bean.getServiceName())));
+
+            server.serviceAt(bean.getPath(), service);
```

Comment:
What happens if the same path is specified multiple times? Will it fail at ```serviceAt```, ```build```, or not fail at all?

In any case, consider validating the paths are unique here to provide a better error message.